File: xocr.py
#!/usr/bin/env python3

from apple_ocr.ocr import OCR
from PIL import Image
import argparse

def get_lines(elements):
    line_threshold = 0.015
    lines = []
    current_line = []
    prev_y = None

    for word, x, y, w in elements:
        if prev_y is None or abs(y - prev_y) < line_threshold:
            current_line.append((word, x, w))
        else:
            lines.append(current_line)
            current_line = [(word, x, w)]
        prev_y = y

    if current_line:
        lines.append(current_line)
    return lines

def correct_lines(lines, min_x):
    GRANULARITY = 0.02
    INDENT_LEVEL = 4
    WORD_SPACING = 0.01
    final_lines = []
    for line in lines:
        line = sorted(line, key=lambda e: e[1])
        first_x = line[0][1]
        relative_x = max(0, first_x - min_x)
        indent_level = int(relative_x / GRANULARITY)
        indent = " " * (indent_level * INDENT_LEVEL)

        line_str = indent
        last_end = line[0][1]

        for i, (word, x, w) in enumerate(line):
            if i == 0:
                line_str += word
            else:
                gap = x - last_end
                num_spaces = max(1, int(gap / WORD_SPACING))
                line_str += " " * num_spaces + word
            last_end = x + w

        final_lines.append(line_str)
    return final_lines

def extract_code(code_source):
    image = Image.open(code_source)
    ocr_instance = OCR(image=image)
    ocr_result = ocr_instance.recognize()

    words = ocr_result['Content']
    xs = ocr_result['x']
    ys = ocr_result['y']
    widths = ocr_result['Length']

    elements = list(zip(words, xs, ys, widths))
    elements.sort(key=lambda e: -e[2])  # sort by y-values

    lines = get_lines(elements)
    min_x = min(e[1] for e in elements)
    corrected_lines = correct_lines(lines, min_x)

    return "\n".join(corrected_lines)

def main():
    parser = argparse.ArgumentParser(description="Accept a file path as input.")
    parser.add_argument("file_path", type=str, help="Path to the input file")
    args = parser.parse_args()

    code = extract_code(args.file_path)
    print(code)

if __name__ == '__main__':
    main()

File: xocr.swift
import Foundation
import Vision
import AppKit

struct OCRWord {
    let text: String
    let x: CGFloat
    let y: CGFloat
    let width: CGFloat
}

func getLines(from elements: [OCRWord], lineThreshold: CGFloat = 0.015) -> [[OCRWord]] {
    var lines: [[OCRWord]] = []
    var currentLine: [OCRWord] = []
    var prevY: CGFloat?

    for word in elements {
        if let prev = prevY, abs(word.y - prev) >= lineThreshold {
            lines.append(currentLine)
            currentLine = [word]
        } else {
            currentLine.append(word)
        }
        prevY = word.y
    }

    if !currentLine.isEmpty {
        lines.append(currentLine)
    }

    return lines
}

func correctLines(_ lines: [[OCRWord]], minX: CGFloat) -> [String] {
    let GRANULARITY: CGFloat = 0.05
    let INDENT_LEVEL = 2
    let WORD_SPACING: CGFloat = 0.05
    var finalLines: [String] = []

    for line in lines {
        let sortedLine = line.sorted { $0.x < $1.x }
        guard let first = sortedLine.first else { continue }

        let relativeX = max(0, first.x - minX)
        let indentLevel = Int(relativeX / GRANULARITY)
        var lineStr = String(repeating: " ", count: indentLevel * INDENT_LEVEL)

        var lastEnd = first.x

        for (i, word) in sortedLine.enumerated() {
            if i == 0 {
                lineStr += word.text
            } else {
                let gap = word.x - lastEnd
                let numSpaces = max(1, Int(gap / WORD_SPACING))
                lineStr += String(repeating: " ", count: numSpaces) + word.text
            }
            lastEnd = word.x + word.width
        }

        finalLines.append(lineStr)
    }

    return finalLines
}

func extractCode(from imagePath: String) {
    let url = URL(fileURLWithPath: imagePath)
    guard let image = NSImage(contentsOf: url),
          let tiffData = image.tiffRepresentation,
          let bitmap = NSBitmapImageRep(data: tiffData) else {
        print("Failed to load image.")
        return
    }

    let request = VNRecognizeTextRequest { request, error in
        guard let observations = request.results as? [VNRecognizedTextObservation] else {
            print("No text found.")
            return
        }

        let words: [OCRWord] = observations.compactMap { observation in
            guard let topCandidate = observation.topCandidates(1).first else { return nil }

            let box = observation.boundingBox
            return OCRWord(
                text: topCandidate.string,
                x: box.origin.x,
                y: box.origin.y,
                width: box.size.width
            )
        }

        let sorted = words.sorted { $0.y > $1.y }
        guard let minX = sorted.map(\.x).min() else { return }

        let lines = getLines(from: sorted)
        let corrected = correctLines(lines, minX: minX)
        print(corrected.joined(separator: "\n"))
    }

    request.recognitionLevel = .accurate
    request.usesLanguageCorrection = true

    let handler = VNImageRequestHandler(cgImage: bitmap.cgImage!, options: [:])
    try? handler.perform([request])
}

// MARK: - Main
let args = CommandLine.arguments

guard args.count > 1 else {
    print("Input Image not found")
    exit(1)
}

extractCode(from: args[1])
